src/index/ivf/ivf.cc

```diff
@@ -270,73 +272,153 @@ IvfIndexNode<T>::Train(const DataSet& dataset, const Config& cfg) {
     auto dim = dataset.GetDim();
     auto data = dataset.GetTensor();
 
-    typename QuantizerT<T>::type* qzr = nullptr;
-    faiss::IndexIVFPQFastScan* base_index = nullptr;
     std::unique_ptr<T> index;
+    // if cfg.use_elkan is used, then we'll use a temporary instance of
+    //  IndexFlatElkan for the training.
     try {
         if constexpr (std::is_same<faiss::IndexIVFFlat, T>::value) {
             const IvfFlatConfig& ivf_flat_cfg = static_cast<const IvfFlatConfig&>(cfg);
             auto nlist = MatchNlist(rows, ivf_flat_cfg.nlist.value());
-            qzr = new (std::nothrow) typename QuantizerT<T>::type(dim, metric.value());
-            index = std::make_unique<faiss::IndexIVFFlat>(qzr, dim, nlist, metric.value(), is_cosine);
+
+            const bool use_elkan = ivf_flat_cfg.use_elkan.value_or(true);
```

Comment:
Just a discussion. Should we put this `use_elkan` inside Faiss or in Knowhere. 
Now we put it in Knowhere:
 - **Pros:** It makes Faiss cleaner, and easier to upgrade next time.
 - **Cons:** As we talked, if we want to make a pure Faiss support in future, we need to keep Knowhere as lightweight as possible. And anyway this PR introduced an `IndexFlatElkan` into Faiss.

Reply:
The way it is introduced satisfies both imho. It makes Faiss cleaner, because it is a single Independent file/index. And it keeps Knowhere lightweight.

Reply:
Aha, we can discuss this when we really need to implement the pure Faiss support. It need Knowhere to be extremely light weight, which means no Faiss' config should be exposed to Knowhere.
Anyway let's keep it for now.